python/pyspark/sql/dataframe.py

```diff
@@ -2044,7 +2044,7 @@ def crossJoin(self, other: "DataFrame") -> "DataFrame":
     def join(
         self,
         other: "DataFrame",
-        on: Optional[Union[str, List[str], Column, List[Column]]] = None,
+        on: Optional[Union[str, List[str], bool, List[bool], Column, List[Column]]] = None,
```

Comment:
That's actually an error from IDE (which assumes that the built-in functions always return `bool`). The expected types here are correct in fact.

Reply:
Yes.

Reply:
I see. I found something interesting.

Intellij IDEA has no issue with pandas `join`'s `on`.
<img width="1461" alt="image" src="https://user-images.githubusercontent.com/3375461/195002907-e951789e-f723-4962-a02b-0abb55698146.png">
But seems only has issue for pyspark.

Here is the way pandas implement:

- https://github.com/pandas-dev/pandas/blob/main/pandas/core/frame.py#L9837-L9840
- https://github.com/pandas-dev/pandas/blob/main/pandas/_typing.py#L121

Reply:
Okay, that's interesting. cc @zero323 fyi.

Reply:
@HyukjinKwon  As far as I can tell, there is no real difference here. The difference between Pandas and Spark check, is most likely related to the missing stubs for the former one. If I use environment without `pandas-stubs` things type check in PyCharm 

![without pandas stubs](https://user-images.githubusercontent.com/1554276/204131753-eb818c1b-9c35-4069-9fd4-abca28f0d57b.gif)

If I choose one with `pandas-stubs` installed, I get 

![with pandas stubs](https://user-images.githubusercontent.com/1554276/204131767-c3f103c4-b8a1-4554-a1a7-723d034abd1b.gif)

which is the same category of failure as for PySpark code.

Given mypy as a reference, this is an expected false positive (see https://github.com/python/mypy/issues/2783).

On a side note ‒ Pandas and PySpark joins shown on the screenshots are not even remotely equivalent.